### PR TITLE
0.5.1 k schedulers

### PR DESCRIPTION
**Pull request by tonetechnician**
@patil-suraj This PR is related to conversions that have been happening for this PR #636.

It ports @hlky's implementation of k-diffusion schedulers up to 0.5.1 of diffusers. 

Currently the output is working for txt2img pipeline (although not matching results from k-diffusion output), but there are still issues with img2img pipeline output which I'm resolving and some of that can be followed in this issue report #901. 

Also some further work needs to be done for DPM2 and Heun solvers which are meant to accept the model and use that to determine the next sample step https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py#L114. I believe @hlky implemented an approximation of this which is why we currently do get reasonable output. 

I have some code that I'm wanting to push up soon that allows the model to passed to the step scheduler, but still solving some issues there.

### CODE REVIEW

**Comment by tonetechnician** [comment]
Here is some example output that I've got with settings:

prompt: "A dense Cyberpunk city street" 
num_inference_steps: 20
guidance_scale: 7.5
seed: 0

EULER:

output-euler

EULER-A:

output-euler-a

DPM2:

output-dpm2

DPM2-A:

output-dpm2-a

HEUN:

output-heun

LMS:

output-lms

PNDM:

output-pndm

DDIM:

output-ddim

DDPM:

output-ddpm

**Comment by andrea-gatto** [comment]
Thanks a lot @tonetechnician for your work!
JFYI I tested your branch with the euler schedulers and I wasn't able to replicate the images above. Also the images produced with these two schedulers look more grainy wrt the ones generated using @hlky ones

**Comment by tonetechnician** [comment]
Thanks for checking it out @andrea-gatto 

There does seem to be a difference between 0.3.0 and 0.4.0 as mentioned here #902 which could be related to what you're experiencing. 

That being said, you should be getting the same output as me, so something is off. I'll dig in a bit more and see what that might be. Would you be able to post your output for reference? 

Thanks!

**Comment by andrea-gatto** [comment]
Thanks @tonetechnician for the heads up
Moving to 0.3.0 doesn't have any effect on the quality for me. 
As you can see below my generations are similar to yours (especially Euler), but with much lower quality.
This is the code I've used to replicate your results

```
import torch
from diffusers import StableDiffusionPipeline, EulerDiscreteScheduler, EulerAncestralDiscreteScheduler

device = "cuda"
model_id = "CompVis/stable-diffusion-v1-4"
scheduler = EulerAncestralDiscreteScheduler()
pipe = StableDiffusionPipeline.from_pretrained(model_id, use_auth_token=True, scheduler=scheduler)
pipe = pipe.to(device)
generator = torch.Generator(device).manual_seed(0)

prompt = "A dense Cyberpunk city street"
with torch.autocast("cuda"):
    image = pipe(prompt, guidance_scale=7.5, num_inference_steps=20, generator=generator).images[0]

image.save("test.png")
```

EULER
euler

EULER-A
euler_ancestral

**Comment by tonetechnician** [comment]
Thanks for reference to images and code @andrea-gatto.

Sorry, I should've posted my scheduler settings:

```
scheduler = EulerAncestralDiscreteScheduler(
    beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear"
)
self.pipe.scheduler = scheduler
output = self.pipe([prompt], generator=generator, guidance_scale=guidance_scale, num_inference_steps=num_inference_steps, seed=seed)
```

Might be that the scheduler is defaulting to linear in yours?

**Comment by andrea-gatto** [comment]
ha I see, I thought you were using the default values
updating your values I'm able to replicate your euler result, still different for euler-a (though the quality has improved dramatically)
EULER
euler
EULER-A
euler_ancestral

**Comment by sausax** [comment]
Great work @tonetechnician . One thing I noticed that this takes little more memory than the main branch. I was getting CUDA memory error on a 8gb card. Worked fine on more memory card.

**Review by patil-suraj — COMMENTED** [review]
Amazing work @tonetechnician ! I would suggest to split this PR into two.
1. for euler schedulers 
2. for DPM and Heun 

The euler schedulers are already working great and would be awesome to have them merged soon! Then we can start discussing the design for DPM & Heun in other PR to so this won't get blocked by it.

**Comment by patil-suraj** [comment]
Hey @tonetechnician , let us know if we could help in any way, would be cool to have the euler schedulers  merged today!

**Comment by tonetechnician** [comment]
Hey @patil-suraj 

All good for the meanwhile, just had some backlog work that prevented me from splitting this out. Should be able to do so today and make a PR

**Comment by patil-suraj** [comment]
No worries ! And awesome, looking forward to it, this is going to be really nice addition!

**Comment by hlky** [comment]
@patil-suraj #1019

**Comment by patil-suraj** [comment]
Thank you @hlky will take a look soon :)

**Comment by tonetechnician** [comment]
Solved here --> https://github.com/huggingface/diffusers/pull/1019

**Comment by patrickvonplaten** [comment]
Should we open a new PR for the 2nd order schedulers? We're still definitely be interested in adding those if you'd like :-)

**Comment by patrickvonplaten** [comment]
I think all the 2nd order schedulers have now been added. Thanks a mille for you first draft here @tonetechnician - it has been super useful to get the other schedulers correctly added.

**Comment by tonetechnician** [comment]
Great work everyone! SO EPIC to see this + 2.0 + depth2img. Unfortunately been MIA on another project that has taken my attention for the last few weeks but back in today :) Pulled latest changes and was happy to see the second order samplers! 

Thanks @patrickvonplaten for the mention! But great work to everyone for getting this in, so good :) Look forward to checking them all out
